Ongaku/ScrobblerController.swift

```diff
@@ -0,0 +1,307 @@
+//
+//  ScrobbleController.swift
+//  Ongaku
+//
+//  Created by Kot on 8/16/23.
+//  Copyright © 2023 Spotlight Deveaux. All rights reserved.
+//
+
+import Foundation
+import Combine
+import CryptoKit
+import AppKit
+import Alamofire
+import KeychainAccess
+import os.log
+
+private let log: Logger = .init(subsystem: "io.github.spotlightishere.Ongaku", category: "view-controller")
+
+class ScrobblerController: ObservableObject {
+	private let player: Player
+	private var playerSink: AnyCancellable?
+	@Published var enabled: Bool = true
+	
+	private var authToken: String?
+	@Published var session: LastFMSession?
+	private var latestScrobbledTrack: ScrobbleData?
+	
+	// What you gonna do, scrobble me to death?
+	let baseUrl = URL(string: "https://ws.audioscrobbler.com/2.0")
+	let authUrl = URL(string: "https://last.fm/api/auth")
+	private let apiKey = "50ad35bfe1bb89b96f0c6e23c724dd5e"
+	private let sharedSecret = "d2673496334bffaf9d5a1fac6bd29887"
+	
+	init(player: Player) {
+		self.player = player
+		playerSink = player.state.sink { state in
+			Task(priority: .userInitiated) {
+				await self.updateScrobbler(playerState: state)
+			}
+		}
+		
+		_ = loadSessionFromKeychain()
+	}
+	
+	func updateScrobbler(playerState state: PlayerState) async {
+		if session == nil && authToken != nil {
+			log.debug("Attempting to authorize Last.fm session")
+			await fetchAndSaveSession()
+		} else {
+			if !enabled {
+				log.debug("Scrobbler is disabled, skipping update")
+				return
+			}
+			
+			switch state {
+			case .playing(let active):
+				let data = ScrobbleData(artist: active.track.artist ?? "Unknown Artist", track: active.track.title, album: active.track.album, duration: Int(active.track.duration))
+				updateNowPlaying(data: data)
+				
+				// Don't scrobble this track if it's the latest scrobbled track
+				let isLastScrobbled = data == latestScrobbledTrack
+				if !isLastScrobbled {
+					if (active.position > active.track.duration / 2 || active.position > 4 * 60) {
+						let ts = Date().timeIntervalSince1970 - active.position
+						sendScrobble(data: data, timestamp: Int(ts))
+					}
+				} else {
+					log.debug("Not scrobbling \(data.artist) - \(data.track) as it was already scrobbled")
+				}
+			default: break
+			}
+		}
+	}
+	
+	func updateNowPlaying(data: ScrobbleData) {
+		if let session {
+			let key = session.key
+			let durationStr = if let duration = data.duration { String(duration) } else { "" }
+			
+			// TODO: maybe move to ScrobbleData class? and use json post body
```

Comment:
The docs imply that it's [preferred to POST](https://www.last.fm/api/show/track.updateNowPlaying) on this route - I see that a few other functions here are similar, while others are GET-only with query parameters. Would it be possible to go ahead and implement this prior to merging?

It seems like this is currently functional so hopefully their API won't change to ever exclude this, but it'd be nice to have it as documented just for future-proofing!

Reply:
Unfortunately Swift's Dictionary doesn't support ordered items, which Last.fm requires to have in request bodies (including POSTed JSON ones).
I tried OrderedDictionary with AlamoFire to see if that would encode in-order, but OrderedDictionary is only encodable to an array, due to being a sorted type. Keyed encoders like JSONEncoder or URLFormParamEncoder aren't supported.

Good news though, plenty if not all of other Swift impls just use query params :P
https://sourcegraph.com/search?q=context:global+file:%5C.swift%24+%22api_sig%22&patternType=standard&case=yes&sm=1&groupBy=repo
https://sourcegraph.com/github.com/Andr3y-/EyePatchPlayer/-/blob/EyePatchPlayer/EPHTTPLastFMManager.swift
https://sourcegraph.com/github.com/heyvito/audioscrobbler/-/blob/Audioscrobbler/WebService.swift
https://sourcegraph.com/github.com/nrubin29/finale/-/blob/ios/FinaleWidget/Lastfm.swift

Unsure if it would be somehow possible with URLSession or something besides Alamofire, but unlikely since the Dictionary isn't sorted, and JSONEncoder itself doesn't work with OrderedDictionary.

Reply:
Gotcha - thank you so much for taking a look! That definitely sounds like it'd be far more hassle than it'd be worth to properly implement :)